### PR TITLE
Fast jwt errors thrown

### PR DESCRIPTION
**Pull request by wbrown766**
#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
After looking at the change and the tests, I think it's significantly more transparent to keep method throwing exceptions by default.

However, I can see the reasoning why you might want them to be hidden - but that should be a developer choice.

Could you put this behavior behind   an option, disabled by default?

**Comment by wbrown766** [comment]
@mcollina thats what I was thinking.
I had some time to look at it today, and have a few questons I'd like a bit of feed back on.

I think it is obvious that we would not want to catch all errors, e.g. "RSA Signatures set as Algorithm in the options require a private and public key to be set as the secret" as these are a breaking errors, but only catch the errors relating to the verify process.
Question 1: Does this sound reasonable, or should a different subset of errors or all errors be returned?

Question 2: I am also assuming that we would not want to alter the request[decoratorName]. Corect?

Question 3A: What would be prefered if false logic or if true logic? i.e.
```javascript
fastify.register(jwt, {
  secret: 'foobar',
  throwErrors: false, // default to true
})
```
or
```javascript
fastify.register(jwt, {
  secret: 'foobar',
  returnErrors: true, // default to false
})
```
Question 3B: Is there something better to call this property?

I was thinking of returning something like
```javascript
// token is valed
{
  [decoratorName]: user
}

// token is invalid
{
  error: originalFastifyError
}
```
This would allow responce.send(jwt.error) to function exactly the same way as if fastify-jwt had thrown the error, and would not be that diferent than calling request[decoratorName] on success.
Question 4: What do you think?

**Comment by mcollina** [comment]
1. if we are swallowing errors, let's do it for all. Make sure you log the error that you swallow.
2. I don't understand the question.
3. I would use the `swallowErrors` option instead, defaulted to `false.`

**Comment by wbrown766** [comment]
I wasn't thinking of swallowing errors.
Current behaviour is to immediately return the appropriate HTTP responce via fastify-error. I was thinking of just giving the developer an easy way to intercept.

Like this.
```javascript
const verified = fastify.jwt.verify(token);
if(verified.error) {
  // do someting
  responce.send(error) // returns HTTP error responce
} esle {
  verified.user // would be the decoded token in the user format
}
```

**Comment by mcollina** [comment]
> I wasn't thinking of swallowing errors.

This is swallowing errors:
https://github.com/fastify/fastify-jwt/pull/265/files#diff-5f176ae2a303e20b8a4c3c5c5073bd23f5e3c206cd113e4c2c43d21aaf1dd3c5R202-R203

**Comment by wbrown766** [comment]
Right, so I didn't have any time last week.

Taking a closer look I found out that yes I am swallowing errors and not as I thought passing them to a different function to handle.

I also found out that the tests are all over the place and not all possibilities are tested, leading me to believe that I had solved the problem as it passed all the tests.

So as I am trying to write better tests to get something to work against, I just end up with more questions.

Are the server methods meant to be semi-interchangeable with the request/reply methods?
I mean bar the obvious things like, a server method would need to be given a token as it can not get it from the request headers, should a server method take the same callbacks and options, perform the same checks and output the same errors as its request/reply counterparts? As they don't seem to.

```javascript
const fastifyJwt = require("./jwt");

const fastify = require("fastify")({
  logger: false
});

const pluginConfig = {
  trusted: (request, { jti }) => {
    jti !== "untrusted";
  },
  secret: "Hello world!",
  decode: {
    complete: false
  },
  sign: {
    algorithm: "HS512",
    expiresIn: "1m",
    noTimestamp: false
  },
  verify: {
    algorithms: "HS512",
    complete: false
  }
};

fastify.register(fastifyJwt, pluginConfig);

fastify.get("/serverVerify", { method: "GET" }, async (req, res) => {
  const token = req.headers.authorization.split(" ")[1];
  const jwt = await fastify.jwt.verify(token);
  res.send({ status: "serverVerify", jwt });
  return res;
});

fastify.get("/requestVerify", { method: "GET" }, async (req, res) => {
  const jwt = await req.jwtVerify();
  res.send({ status: "requestVerify", jwt });
  return res;
});

async function start() {
  try {
    await fastify.ready();
    await fastify.listen({ port: 3000, host: "127.0.0.1" });
  } catch (error) {
    fastify.log.error(error);
    process.exit(1);
  }
}
start();

async function serverVerify(jwt) {
  return await fastify
    .inject({
      method: "GET",
      url: "/serverVerify",
      headers: {
        Authorization:
          "Bearer eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJ1c2VySUQiOiIwMDAwMDAwMC0wMDAwLTAwMDAtMDAwMC0wMDAwMDAwMDAwMDAiLCJpYXQiOjEwMDAwMDAwMDAsImV4cCI6MTAwMDAwMDAwMDAsImp0aSI6InVudHJ1c3RlZCJ9.m8T9vYu0dgfTT4i19QoN1bpC18cO68iFmHruoMjZM79IH4FlTrWhHtxDq_yXo6cG1-FJRmNY-wOcAnm0SLoBSg"
      }
    })
    .then((response) => {
      console.log({ responseStatuscode: response.statusCode, body: response.body });
    })
    .catch((error) => {
      console.log({ responseStatuscode: error.statusCode, body: error.body });
    });
}

async function requestVerify(jwt) {
  return await fastify
    .inject({
      method: "GET",
      url: "/requestVerify",
      headers: {
        Authorization:
          "Bearer eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJ1c2VySUQiOiIwMDAwMDAwMC0wMDAwLTAwMDAtMDAwMC0wMDAwMDAwMDAwMDAiLCJpYXQiOjEwMDAwMDAwMDAsImV4cCI6MTAwMDAwMDAwMDAsImp0aSI6InVudHJ1c3RlZCJ9.m8T9vYu0dgfTT4i19QoN1bpC18cO68iFmHruoMjZM79IH4FlTrWhHtxDq_yXo6cG1-FJRmNY-wOcAnm0SLoBSg"
      }
    })
    .then((response) => {
      console.log({ responseStatuscode: response.statusCode, body: response.body });
    })
    .catch((error) => {
      console.log({ responseStatuscode: error.statusCode, body: error.body });
    });
}

serverVerify();
requestVerify();
```
server method does not care if the token is trusted or not
the trusted function is never run
```json
{
  "responseStatuscode": 200,
  "body": {
    "status": "serverVerify",
    "jwt": { "userID": "00000000-0000-0000-0000-000000000000", "iat": 1000000000, "exp": 10000000000, "jti": "untrusted" }
  }
}
```
the request method however does run the trusted function and does return an error
```json
{
  "responseStatuscode": 401,
  "body": {
    "statusCode": 401,
    "code": "FST_JWT_AUTHORIZATION_TOKEN_UNTRUSTED",
    "error": "Unauthorized",
    "message": "Untrusted authorization token"
  }
}
```
This is not the only inconsistency, but I am limited in the amount of time I can spend and feel this demos the contrast quickly and clearly.

File: jwt.js [hunk 1]
```diff
@@ -190,14 +190,18 @@ function fastifyJwt (fastify, options, next) {
...
+      if (options && typeof options !== 'function') {
+        const localDecoder = createDecoder(options)
+        return localDecoder(token)
+      }
 
-    return decoder(token)
+      return decoder(token)
+    } catch (err) {
+      return undefined
+    }
```
Comment (jsumners) [inline jwt.js:204]: I don't think returning `undefined` is a good substitute to indicate there was an issue. Errors should be checked. I would recommend returning a monadic object like:

```js
{
  error: Error('decode failed for a specific reason'),
  result: 'set when there was not an error'
}
```

**Comment by mcollina** [comment]
My understanding is that this can be closed. If I'm wrong, let me know!

**Pull request closed without merging**